[Bug 14655] com.livecode.foreign: Add "UIntSize" type.

peter-b opened the pull request:
Add a foreign type to LCB corresponding to a memory offset (C's `size_t`).

peter-b committed:
libfoundation: Fix up and extend integer hashing API.

* Add templates for hashing arbitrary-sized signed/unsigned integers
  (this fixes a minor bug where the hash of a pointer on 64-bit
  platforms would be independent of the upper 32-bits of the pointer
  if hash_t was defined to be 32-bit).

* Add MCHashUInteger() and MCHashUSize() functions.

peter-b committed:
libfoundation: Add foreign size type info structure.

Currently, this will only successfully bridge to an MCNumberRef iff it
can fit within the MCNumberRef's unsigned range (i.e. 32 bits).

peter-b committed:
com.livecode.foreign: Add "UIntSize" type.

We can't use "size" because "size" is a keyword already.

runrevmark commented on libfoundation/src/foundation-core.cpp:
```diff
@@ -205,9 +205,17 @@ void MCMemoryDeleteArray(void *p_array)
 #define HASHFACTOR 2654435761U
 
-hash_t MCHashInteger(integer_t i)
+hash_t MCHashInteger(int64_t i)
```
Is this the 'right' thing to do?

'hash_t' can change in size (although there might be an argument that it should remain 32-bit given what hash's tend to be used for)

Wouldn't it be better to have 'sized' hash functions? HashInt32 / HashUInt32 / HashInt64 / HashUInt64

peter-b commented:
@runrevmark I think that it makes sense for `MCHashInteger()` and `MCHashUInteger()` to take arguments of the largest integer size we use.  The reason for this is that C/C++ will correctly promote smaller integer arguments.

At the moment, `hash_t` is platform-word-sized. So this patch isn't correct :sob: , because it will probably misbehave on platforms where `hash_t` is `uint32_t` (i.e. 32-bit platforms).

Where did you borrow this hash algorithm from? Do you recall what it did with variously-sized integers?

runrevmark commented:
Hmmm - I think hash_t should be 32-bit at the moment regardless (we should be MEDIUM memory model, not LARGE - although the world hasn't stopped so I'm guessing this is okay - modulo namerefs taking extra space).

The reason I think there should be separate int-sized hash functions is efficiency.

Obviously Hash(int) should produce the same hash for the same value regardless of the bit-width of int. However, if hash_t is 32-bit, then it would make sense that computing Hash(32-bit int) should be faster than computing Hash(64-bit int).

By making the interface 'always promote' you take away both the choices of implementation for the hash (you have to always use a 64-bit hash algorithm and then restrict to 32-bit), and make 32-bit platforms do more work (which they really cannot afford to do - think cheap Android devices on ARM).

runrevmark commented:
I got the hash from CoreFoundation:

http://stackoverflow.com/questions/5194781/how-to-handle-floats-and-doubles-when-implementing-hash-in-objective-c

(which itself seems to come from Knuth :))

peter-b committed:
libfoundation: Use MCHashUInteger() for hashing foreign uint values.

peter-b commented:
@runrevmark I've fixed up the changeset based on your suggestions:
- I've added generic size-independent `MCHashInt()` and `MCHashUInt()` templates, and redirected existing integer hash functions to use them.
- I've added `MCHashUInteger()` and `MCHashUSize()` functions.

Hopefully since the templates are marked as inline, they'll optimise away pretty nicely.

peter-b committed:
Merge branch 'develop' into bugfix-14655

Conflicts:
	libscript/src/foreign.mlc

peter-b commented:
@runrevmark Any chance of another review please?

runrevmark commented on libfoundation/src/foundation-foreign.cpp:
```diff
@@ -280,6 +297,18 @@ static bool __uint_import(void *contents, bool release, MCValueRef& r_value)
+{
+	size_t t_value = *(size_t *) contents;
+	if (t_value > UINTEGER_MAX)
+		return false;
```
I think you need to throw an error here - otherwise you won't see why the mapping failed at the script-level.

peter-b committed:
libfoundation: Throw an error when imported size_t is too large for uinteger_t

peter-b commented:
@runrevmark Importing a `size_t` now properly throws an error when too large.

runrevmark commented:
Then you may have a reviewed label :)

Reviewed-by: @runrevmark.
